### PR TITLE
APV pulse shape in a separated python file

### PR DESCRIPTION
**Pull request by bourgatte**
The APV pulse shape is now in a separated file and then stored in a vector whatever its size to easily modify it.
Assuming this is purely technical, nothing should change.

### CODE REVIEW

**Comment by makortel** [comment]
Moving the ~1.5k numbers as doubles to be the configuration is not necessarily the best choice (from the framework point of view). Assuming the plan is to allow changes in the numbers, could they be moved e.g. to EventSetup, or to a separate text file pointed to with `cms.FileInPath`?

**Comment by mmusich** [comment]
Moving to ES, with corresponding need to define a record (and eventually a `CondFormat`) would give indeed the maximum flexibility though probably an overkill for something we might want to change with not more than yearly or multi-yearly frequency. OTOH one might want to store the parameters in an era dependent way (e.g. to emulate pre- and post- VFP APV parameters change). If `FileInPath` mechanism allows for such flexibility (i.e. multiple input files in the data repo steered via era), it sounds a good option.

**Comment by makortel** [comment]
@mmusich

> If FileInPath mechanism allows for such flexibility (i.e. multiple input files in the data repo steered via era), it sounds a good option.

Sure, it's a configuration parameter type as others, so can be customized with `Modifier`s (eras). That could then be a good way to proceed (if there indeed is no plan to change these values e.g. within run-dependent MC).

**Comment by bourgatte** [comment]
I used instead a text file using cms.FileInPath and I will commit to update this pull request.

File: SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.cc [hunk 1]
```diff
@@ -19,8 +19,32 @@ SiLinearChargeDivider::SiLinearChargeDivider(const edm::ParameterSet& conf) :
...
+  
+{
+  std::ifstream APVShapeReadFile(peakMode?(conf.getParameter<edm::FileInPath>("APVShapePeakFile").fullPath().c_str()):(conf.getParameter<edm::FileInPath>("APVShapeDecoFile").fullPath().c_str()));
+  if (!APVShapeReadFile.good()){ 
+    if(peakMode){
+      edm::LogError("FileError") << "Problem opening APV Shape file: "<<conf.getParameter<edm::FileInPath>("APVShapePeakFile").fullPath();
+      return;
+    }
+    else {
+      edm::LogError("FileError") << "Problem opening APV Shape file: "<<conf.getParameter<edm::FileInPath>("APVShapeDecoFile").fullPath();
```
Comment (makortel) [inline SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.cc]: Would it make sense to throw an exception instead? (or can one do a meaningful simulation without this information?)

File: SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.cc [hunk 2]
```diff
@@ -19,8 +19,32 @@ SiLinearChargeDivider::SiLinearChargeDivider(const edm::ParameterSet& conf) :
...
+      return;
+    }
+  }
+  while(!APVShapeReadFile.eof()){
+    APVShapeReadFile>>value;
+    valuesVector.push_back(value);
+  }
+  t0Idx=std::distance(valuesVector.begin(),std::find(valuesVector.begin(), valuesVector.end(), 1));
+  if(t0Idx==valuesVector.size()){
+    edm::LogError("WrongAPVPulseShape")<<"The max value of the APV pulse shape stored in the text file used in SimGeneral/MixingModule/python/SiStripSimParameters_cfi.py is not equal to 1. Need to be fixed.";
```
Comment (makortel) [inline SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.cc]: Exception here as well?

File: SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.h [hunk 3]
```diff
@@ -13,6 +13,9 @@
 
 #include "SimGeneral/HepPDTRecord/interface/ParticleDataTable.h"
 
+#include <fstream>
+#include <string>
```
Comment (makortel) [inline SimTracker/SiStripDigitizer/plugins/SiLinearChargeDivider.h]: Please move these includes to the `.cc` as there is nothing in the header requiring them.

**Comment by bourgatte** [comment]
I modified what you suggested, moved resolution into pulse shape file and reduce peak&deco duplication in SiLinearChargeDivider. I will commit.

**Comment by pieterdavid** [comment]
I think that the last commit by @bourgatte addressed all comments that were made so far... could this PR be tested and reviewed again? Thanks!

**Comment by fabiocos** [comment]
please test

**Comment by civanch** [comment]
@bourgatte , are these new files unchanged with time? How many such files you foresee in few years? In other cases, data files are moved out of git and are supported as externals to CMSSW.

Also I see few places of source code with fabs(), which should substituted by std::abs()

**Comment by bourgatte** [comment]
@civanch, the original purpose of this PR was to remove the harcoded values and make them external and configurable. The current values are still similar to the original hardcoded ones. In the mid term we foresee to change them by update ones. We may have in few years few files corresponding to several periods like run II and run III but not really more. We thus don't plan to use externals such as DB. About "fabs" function, I didn't introduce any call in my PR but I can modify the files if needed. Let me know.

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@bourgatte @civanch the files containing the shapes look relatively small to me, but I am a bit surprised by their location: are these supposed to be temporary tests? I would say no, looking at the code and at the review. I would naively imagine that SimTracker/SiStripDigitizer/data would be a more appropriate place for these txt files

**Comment by bourgatte** [comment]
@fabiocos there is no data directory in SimTracker/SiStripDigitizer or in SimGeneral/MixingModule. Do I have to create it and commit? Also currently text files are in SimGeneral/MixingModule, do I have to move them to the other you suggested which is SimTracker/SiStripDigitizer ?

**Comment by fabiocos** [comment]
@bourgatte all the technical code forthe siStrip digitization sits in SimTracker/SiStripDigitizer, apart for the configuration in the SimGeneral/MixingModule, so it would look a cleaner solution to me to move these files into a new data directory there. Alternatively, I see that pixel custom inefficiencies have been provided in https://cmssdt.cern.ch/lxr/source/SimGeneral/MixingModule/python/SiPixelCustomInefficiencyParameters_cfi.py as a python fragment. But I think the former solution is more along the spirit of your PR.

**Comment by fabiocos** [comment]
please test

**Comment by pieterdavid** [comment]
The differences are in the 150.0 workflow, so I suppose they are due to https://github.com/cms-sw/cmssw/issues/25570 (if there were anything wrong with the changes in this PR we should see differences in all workflows that include digitization)

**Comment by civanch** [comment]
+1

**Comment by mmusich** [comment]
@fabiocos, may I kindly suggest you to review this PR and eventually merge it soon? 
The code changes introduced here should not be changing physics at all, i.e. are of purely technical nature.
On the other hand, we have some other changes which rely on top of these ones which are expected to change simulation and we'd really appreciate to get the new PR in by the time of 10_5_0. 
Please let us know,
@pieterdavid @echabert

**Comment by fabiocos** [comment]
@mmusich this looks ok to me to enter in next IB, my comment was taken into account

**Comment by fabiocos** [comment]
+1

**Pull request merged**